### ISSUE TITLE
Recorded By filter not filtered

### ISSUE THREAD

**Issue by gbif-portal** [gbif/portal-feedback#4053]
**Recorded By filter not filtered**

I want to restrict my already filtered occurrence search by a further recordedBy filter. The search there seems to be global though, not including my already existing filters. Especially when I am uncertain about the exact name (Groom, Groom Q. Groom B., ...) this is unfortunate

-----

User: [See in registry](https://www.gbif.org/api/feedback/user/c2992ea2db77cf7d7c65fc67c6a28ec0:527b3bfa539af02488f87e2f3bf1fc39321a809a4b0363a16d409ab49d539150d2ff3fec7e992439d7f8d7d4b4e9ca01f4e410e5de53dbd79c677d5cc60030f19b46d203511eb122bb4cf7fd654ecc25a96e45016eba04b687aebd9f5df15c266fafd45805c7295d212bb68e71f3768ef2cc430c504f9305bcf32ab701bc2622) - [Send email](https://www.gbif.org/api/feedback/user/mailto/c2992ea2db77cf7d7c65fc67c6a28ec0:527b3bfa539af02488f87e2f3bf1fc39321a809a4b0363a16d409ab49d539150d2ff3fec7e992439d7f8d7d4b4e9ca01f4e410e5de53dbd79c677d5cc60030f19b46d203511eb122bb4cf7fd654ecc25a96e45016eba04b687aebd9f5df15c266fafd45805c7295d212bb68e71f3768ef2cc430c504f9305bcf32ab701bc2622)
System: Safari 15.4.0 / Mac OS X 10.15.7
Referer: https://www.gbif.org/occurrence/search?dataset_key=d8cd16ba-bb74-4420-821e-083f2bac17c2&recorded_by=Groom%20Q.&advanced=1
Window size: width 1299 - height 844
[API log](https://private-logs.gbif.org/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-05-02T14:53:44.136Z',to:'2022-05-02T14:59:44.136Z'))&_a=(columns:!(_source),filters:!(),index:'3390a910-fcda-11ea-a9ab-4375f2a9d11c',interval:auto,query:(language:kuery,query:''),sort:!()))
[Site log](https://private-logs.gbif.org/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-05-02T14:53:44.136Z',to:'2022-05-02T14:59:44.136Z'))&_a=(columns:!(_source),filters:!(),index:'5c73f360-fce3-11ea-a9ab-4375f2a9d11c',interval:auto,query:(language:kuery,query:''),sort:!()))
System health at time of feedback: OPERATIONAL

**Comment by MortenHofft** [comment]
@mdoering if I understand correctly you mean that the suggest should be limited to the remaining results/hits right? So if I only have occurrences by `recordedBy:mdoring` left, then it will only show `mdoring` as a suggestion.

I sometimes feel the same, but this of course is not how any of our other suggest filters work either: e.g. taxonKey, country, dataset, publisher, collectionCode, etc. They all suggest within the global scope.

It would be great if we could find a way (UI as well as backend) to support both options. 

@javiesgm you might have a good idea for how to do so in the UI (hosted portal type). 

@mdoering do you have any feeling for how taxonomy suggest driven by ES and the current filter would/could work/perform?

**Comment by mdoering** [comment]
If that is the case I think it is wrong for all of the others too.
All other facet filters do that filtering and I would expect the same from the suggested filters. If you look up the basis of record for the above link it will just show numbers for the INSDC Sequence occurrences and not global results. You want to be able to dig deeper into the data by using the filters, not start from scratch.

**Comment by MortenHofft** [comment]
I'm a bit confused, since it has been like this forever. And is like that for all search. Species search including
https://www.gbif.org/species/search?dataset_key=046bbc50-cae2-47ff-aa43-729fbf53f7c5&highertaxon_key=1&advanced=1

Can you help me how to do a e.g. dataset suggest that is limited to the already filtered results (say I start with an Animalia filter). Isn't this something that requires API changes as well? Or am I misunderstanding the issue?

**Comment by mdoering** [comment]
You are right. It can't be done with what we have now in all areas. All suggests would have to go through the occurrence API then in order to do the filtering. But would that not be sth to strive for in the future? From my user perspective that is how I would like it to work. I guess I just expected the recordedBy to be driven by the occurrence API / Elastic already and therefore that it's handled the same way as we do for other controlled values.

**Comment by MortenHofft** [comment]
Currently it is driven from ES, but an optimized suggest index that does not reflect other filters. Fede knows the details. But on the hosted portals I use a different approach where it could easily be made to work (though the UI might be confusing)

**Comment by sylvain-morin** [comment]
Hi @MortenHofft ,

Did you or your team implement such approach on the hosted portals recently?

I had this idea in mind for several weeks (when we discovered the hosted portals) and I was on the edge to submit it... but I can't reproduce it anymore for "dataset" criteria for example !?

Indeed, on gbif.org portal, I understand it's complex to filter suggestions based on the other selected criteria (I guess it's for performance reason?)

However, for hosted portals, I find confusing to do suggestions on the global data scope...
For example, it's quite strange to get "France" as country suggestion on the North America hosted portal (https://www.gbif-north-america.org/data/) :-) 

It looks like you did something for dataset (I can't find anymore our french datasets for example).
Do you plan  to restrict the suggestions to the hosted portal scope for all criteria??

Thanks - hosted portal is game changer! Congratulations.

**Comment by MortenHofft** [comment]
Hi @sylmorin-gbif and thank you

**Short answer:** yes I will try to restrict suggestions to the data in scope.

**Loooong answer:**
We have a long standing issue here https://github.com/gbif/hosted-portals/issues/187

For GBIF.org as well as for the hosted portals we use our standard autocompletions/suggestion endpoints. Those are fast. But the downside is that they are not filtered - not by the hosted portals data scope nor by the current filter.

This behaviour feels more weird for some filters than for others. We have always suggested checklists (in dataset suggest) without occurrences in GBIF.org for example. And taxon names without occurrences. But for the hosted portals this feels far more weird as the data is more targeted.

I would still argue that suggesting e.g. taxon names without data isn't pointless. In a way it says: we know what you mean, but unfortunately we have no data. But I also get that it is annoying. It would be nice if at least it then showed in the suggest `Puma concolor (no occurrences)` and then perhaps an option to restrict to current result set.

Anyway, it requires some consideration for each filter. And also a fair amount of rewriting for some of them. Country suggest for example is client side as our APIs are not translated. And the browser knows nothing about where occurrences come from. It is all doable of course, but it isn't just one solution I can apply to all fields.

I have started that work. And that is why you will see that Dataset and Publisher search is now restricted to the current portal's data scope (not the current filter). E.g. https://herbarium.gbif.de/data/ will not suggest all datasets in GBIF. But it will still suggest all datasets that are available in that portal even if you have a filter https://herbarium.gbif.de/data/?catalogNumber=139007

The downside is that dataset suggestions are now worse. The set of possible results are better, but the matching isn't as good. Say I want to find the dataset `Danish Mycological Society, fungal records database`. For the proper global dataset suggest I can search for `fungal danish` (reverse order) and get a match. But for the new hosted portals approach the same search will give no results - instead I have to use the correct ordering of the words `danish fungal`.

RecordedBy, IdentifiedBy, StateProvince, Locality, WaterBody, VerbatimScientificName is also scoped by the hosted portal

I hope that makes sense - it is a bit difficult to describe.

**Comment by sylvain-morin** [comment]
Thank for this detailed answer!
It was crystal clear - I can really feel how much work it implies (like migrating from static country list to a country list coming from the data scope, etc)
Good luck for the next developement, it's great to know you are working on it (I will share this info to my colleagues at GBIF France)
